Pre-commit config, linting, and type hints

sondrelg opened the pull request:
Hi @jerry-git 🙂 

I'm going to sit down and have a go at writing the implementation suggested in https://github.com/jerry-git/pytest-split/issues/11 to see if it works. 

I noticed that there was a pre-commit config in the project but that it maybe wasn't fully functional, so I've updated it (let me know if you prefer to remove some of the hooks I've added), run it on the project, and added type hints to the project. Hope that's something you want.

Opening this as an initial PR so it doesn't affect the diff for the next PR(s) 🙂 

If you want, I could also add a [linting workflow](https://github.com/snok/pep585-upgrade/blob/master/.github/workflows/test.yml#L6) and/or test workflow to the project to run on pull-requests

jerry-git commented on setup.cfg:
```diff
@@ -1,2 +1,4 @@
\ No newline at end of file
+max-line-length = 88
+exclude =
+    tests/*
```
what's the motivation behind not linting tests?

sondrelg replied:
We should lint tests, I just had so many errors I originally decided to opt-out, but I can fix them instead 🙂

sondrelg replied:
Added back in now 👍

jerry-git commented on setup.py:
```diff
@@ -25,14 +22,18 @@
         "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
```
Let's add these only when they are tested in the CI

sondrelg replied:
I'll add the workflows to this PR :)

jerry-git commented on src/pytest_split/plugin.py:
```diff
@@ -1,14 +1,23 @@
 from _pytest.config import create_terminal_writer
+from _pytest.main import Session
+
+if TYPE_CHECKING:
```
If we introduce type hints, let's introduce type checking as well. For example, `mypy` pre-commit hook would be nice 🙂

jerry-git replied:
or is `flake8-type-checking` running mypy under the hood?

sondrelg replied:
No it doesn't - I agree, I'll add mypy!

jerry-git commented:
> If you want, I could also add a linting workflow and/or test workflow to the project to run on pull-requests

Sure that would be nice. And it would be actually great if GHA would be used for all the CI related activities.

jerry-git commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,49 @@
+        run: pip install -e .
+      - name: Run tests
+        run: |
+          pytest --cov=pytest_diff --verbose --assert=plain
```
Aren't we now running the tests with both Travis and GHA?

sondrelg replied:
I've never used Travis. How is that configured? Don't see any traces of it in the repo or elsewhere 😛

jerry-git replied:
See https://travis-ci.org/github/jerry-git/pytest-split/branches and https://github.com/jerry-git/pytest-split/blob/master/.travis.yml

jerry-git commented on .pre-commit-config.yaml:
```diff
@@ -1,10 +1,37 @@
+    rev: 'v0.812'
+    hooks:
+      - id: mypy
+        additional_dependencies: [django, djangorestframework, inflection, openapi-spec-validator, prance, pyYAML, django-stubs, djangorestframework-stubs, drf_yasg, drf-spectacular]
```
are you sure django&friends are needed here 😅

sondrelg replied:
I'm sure they're not! 😄

sondrelg commented:
> > If you want, I could also add a linting workflow and/or test workflow to the project to run on pull-requests
> 
> Sure that would be nice. And it would be actually great if GHA would be used for all the CI related activities.

Workflow now seems to work (see [here](https://github.com/sondrelg/pytest-split/actions/runs/892070450)) 🎉 

I removed tests from the flake8-ignore list, added mypy, and added configs for the flake8-packages used 🙂

jerry-git reviewed (approved):
Great stuff 💪 

Thank you!
